grakn-engine/src/main/java/ai/grakn/engine/postprocessing/UpdatingInstanceCountTask.java

```diff
@@ -0,0 +1,117 @@
+import ai.grakn.engine.tasks.TaskCheckpoint;
+import ai.grakn.factory.EngineGraknGraphFactory;
+import ai.grakn.graph.admin.ConceptCache;
+import ai.grakn.util.ErrorMessage;
+import mjson.Json;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * <p>
+ *     Task that controls when types are updated with their new instance counts
+ * </p>
+ *
+ * <p>
+ *     This task begins only if enough time has passed (configurable) since the last time a job was added.
+ * </p>
+ *
+ * @author fppt
+ */
+public class UpdatingInstanceCountTask implements BackgroundTask {
+    private static final Logger LOG = LoggerFactory.getLogger(GraknEngineConfig.LOG_NAME_POSTPROCESSING_DEFAULT);
+    private ConceptCache cache = EngineCacheProvider.getCache();
+
+    @Override
+    public boolean start(Consumer<TaskCheckpoint> saveCheckpoint, Json configuration) {
+        cache.getKeyspaces().parallelStream().forEach(this::updateCountsOnKeySpace);
+        return true;
+    }
+
+    private void updateCountsOnKeySpace(String keyspace){
+        Map<TypeLabel, Long> jobs = new HashMap<>(cache.getInstanceCountJobs(keyspace));
+        //Clear the cache optimistically because we think we going to update successfully
+        jobs.forEach((key, value) -> cache.deleteJobInstanceCount(keyspace, key));
+
+        //TODO: All this boiler plate retry should be moved into a common graph mutating background task
+
+        boolean notDone = true;
+        int retry = 0;
+
+        while(notDone) {
+            notDone = false;
+            try (GraknGraph graknGraph = EngineGraknGraphFactory.getInstance().getGraph(keyspace, GraknTxType.WRITE)) {
+                graknGraph.admin().updateTypeCounts(jobs);
+                graknGraph.admin().commitNoLogs();
+            } catch (Throwable e) {
+                LOG.error("Unable to updating instance counts of graph [" + keyspace + "]", e);
+                if(retry > 10){
+                    LOG.error("Failed 10 times in a row to update the counts of the types on graph [" + keyspace + "] giving up");
+                    jobs.forEach((key, value) -> cache.addJobInstanceCount(keyspace, key, value));
+                } else {
+                    retry = performRetry(retry);
+                    notDone = true;
+                }
+            }
+        }
+    }
+
+    private static int performRetry(int retry){
+        retry ++;
+        double seed = 1.0 + (Math.random() * 5.0);
+        double waitTime = (retry * 2.0)  + seed;
+        LOG.debug(ErrorMessage.BACK_OFF_RETRY.getMessage(waitTime));
+
+        try {
+            Thread.sleep((long) Math.ceil(waitTime * 1000));
+        } catch (InterruptedException e1) {
+            LOG.error("Exception",e1);
+        }
+
+        return retry;
+    }
+
+    @Override
+    public boolean stop() {
+        throw new UnsupportedOperationException("Not yet implemented");
```

Comment:
Maybe we should make these `default` implementations in the interface @alexandraorth